src/vos/vos_obj.c

```diff
@@ -122,12 +174,39 @@ key_punch(struct vos_object *obj, daos_epoch_t epoch, uint32_t pm_ver,
 			rc = key_tree_punch(obj, toh, epoch, &akeys[i], &riov,
 					    flags, ts_set, &dkey_info,
 					    &akey_info);
+			if (rc == -DER_TX_RESTART) {
+				read_conflict = true;
+				rc = 0; /** We should do existence checks on
+					 *  all the akeys first.
+					 */
+			}
+
 			if (rc != 0) {
 				VOS_TX_LOG_FAIL(rc, "Failed to punch akey: rc="
 						DF_RC"\n", DP_RC(rc));
 				break;
 			}
 		}
+
+		if (!read_conflict && rc == 0 && ts_set) {
+			rc = tree_is_empty(obj, toh, &epr, VOS_ITER_AKEY);
+			if (rc > 0) {
+				/** akey tree is now empty, go punch the dkey */
+				key_tree_release(toh, 0);
+				vos_ts_set_punch_propagate(ts_set,
+							   VOS_TS_TYPE_DKEY);
+				propagated = 1;
```

Comment:
The "propagated" will be reset at line 117 after executing "goto punch_dkey", expected?